bonfire/bonfire.py

```diff
@@ -1016,6 +1095,190 @@ def _cmd_apps_what_depends_on(
     print("\n".join(found) or f"no apps depending on {component} found")
 
 
+@reservation.command("create")
+@click.option(
+    '--bot',
+    '-b',
+    is_flag=True,
+    help="Use this flag to skip the duplicate reservation check (for automation)",
+)
+@options(_reservation_process_options)
+@options(_timeout_option)
+def _create_new_reservation(bot, name, requester, duration, timeout):
+    def _err_handler(err):
+        msg = f"reservation failed: {str(err)}"
+        _error(msg)
+
+    try:
+        res = get_reservation(name)
+        # Name should be unique on reservation creation.
+        if res:
+            raise FatalError(
+                f"Reservation with name {name} already exists"
+            )
+
+        res_config = process_reservation(name, requester, duration)
+
+        log.debug("processed reservation:\n%s", res_config)
+
+        if not bot:
+            if check_for_existing_reservation(res_config["items"][0]["spec"]["requester"]):
+                _warn_of_existing(res_config["items"][0]["spec"]["requester"])
+
+        try:
+            res_name = res_config["items"][0]["metadata"]["name"]
+        except (KeyError, IndexError):
+            raise Exception(
+                "error parsing name of Reservation from processed template, "
+                "check Reservation template"
+            )
+
+        apply_config(None, list_resource=res_config)
+
+        ns_name = wait_on_reservation(res_name, timeout)
+    except KeyboardInterrupt as err:
+        log.error("aborted by keyboard interrupt!")
+        _err_handler(err)
+    except TimedOutError as err:
+        log.error("hit timeout error: %s", err)
+        _err_handler(err)
+    except FatalError as err:
+        log.error("hit fatal error: %s", err)
+        _err_handler(err)
+    except Exception as err:
+        log.exception("hit unexpected error!")
+        _err_handler(err)
+    else:
+        log.info(
+            "namespace '%s' is reserved by '%s' for '%s'",
+            ns_name,
+            res_config["items"][0]["spec"]["requester"],
+            duration,
+        )
+        click.echo(ns_name)
+
+
+@reservation.command("extend")
+@options(_reservation_lookup_options)
+@options(_reservation_process_options)
+def _extend_reservation(name, namespace, requester, duration):
```

Comment:
I think we might want some input validation on 'duration'? I was able to pass `4` to it and the command accepted it, but that actually appears to do nothing to the reservation duration. See usage of `_validate_set_image_tag` a bit higher up in this file as an example of using a callback for parameter validation. Is there a regex we can match against for the format we're expecting for 'duration'?

```
❯ bonfire reservation list
2021-09-29 12:04:35 [    INFO] [          MainThread] running (pid 171315): oc get reservation 
2021-09-29 12:04:35 [    INFO] [          pid-171315]  |stdout| NAME                           REQUESTER   READY     NAMESPACE          EXPIRATION
2021-09-29 12:04:35 [    INFO] [          pid-171315]  |stdout| bonfire-reservation-6a60d6c5   bonfire     true      ephemeral-jqacy6   2021-09-29T20:00:41Z
❯ bonfire reservation extend -r bonfire -d 4
2021-09-29 12:04:47 [    INFO] [          MainThread] processing namespace reservation
2021-09-29 12:04:47 [    INFO] [          MainThread] running (pid 171433): oc apply -f - 
2021-09-29 12:04:48 [    INFO] [          pid-171433]  |stdout| namespacereservation.cloud.redhat.com/bonfire-reservation-6a60d6c5 configured
2021-09-29 12:04:48 [    INFO] [          MainThread] reservation 'bonfire-reservation-6a60d6c5' extended by '4'
❯ bonfire reservation list
2021-09-29 12:04:52 [    INFO] [          MainThread] running (pid 171511): oc get reservation 
2021-09-29 12:04:52 [    INFO] [          pid-171511]  |stdout| NAME                           REQUESTER   READY     NAMESPACE          EXPIRATION
2021-09-29 12:04:52 [    INFO] [          pid-171511]  |stdout| bonfire-reservation-6a60d6c5   bonfire     true      ephemeral-jqacy6   2021-09-29T20:00:41Z
```

Reply:
So after reading around a bit more, the duration is a go `time.Duration` ... maybe ensuring our 'duration' matches a regex along these lines will be good enough? https://stackoverflow.com/questions/48765574/regex-how-to-match-string-time-duration-with-optional-conditions

Reply:
Or maybe we could handle the input validation in the CRD spec of the operator somehow instead of here in bonfire and the 'oc apply' would fail. As it stands now though, the bad duration made it through to the operator:

```
2021-09-29T12:04:58.374-0400	ERROR	controller-runtime.manager.controller.namespacereservation	Reconciler error	{"reconciler group": "cloud.redhat.com", "reconciler kind": "NamespaceReservation", "name": "bonfire-reservation-6a60d6c5", "namespace": "", "error": "time: missing unit in duration \"4\""}
```

Reply:
I added a short regex check in the option callback. We could probably do the validation in the operator, but the option callback is pretty easy.

Reply:
I restricted it to XhYmZs, and added an example if an incorrect format is passed in

Reply:
I didn't include support for 'd' (days). We may want to add that in on the front

Reply:
Actually I'm just going to keep it H/M/S instead of days.